File: Remove_nthEnd_Node.cpp
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : no issues as of now.. Learning

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
         ListNode* dummyhead = new ListNode;
        dummyhead->next = head;
       ListNode* p1 = dummyhead; // dummery head for the corner case of single node case.
        ListNode* p2 = dummyhead;
        int counter = 0;
        if (n ==0) return head;
        while(p1!=NULL){
            if (counter > n){
                p2 = p2->next; // as soon as p1 exceeds the n values we start moving the p2
            }
            p1= p1->next;
            counter+=1;
          
        }
        
        p2->next = p2->next->next; // updating the position in order to skip the next node i.e remove it
        
      return dummyhead->next;  // nothing but the head node. but in case of single node and n=1 this will return NULL.
    }
};

File: RemoveNthEndNode.java
// Time Complexity : O(n), traversing in one pass
// Space Complexity : O(1), only using pointers
// Did this code successfully run on Leetcode :yes
// Three line explanation of solution in plain english : Keep two pointers with difference of n between them and move them forward. When fast reaches null, slow will be at n+1th node from end. 
//Remove the next node from the list.


// Your code here along with comments explaining your approach

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(-1);
        dummy.next = head;

        ListNode fast = dummy;
        ListNode slow = dummy;
        int count = 0;

        // keep differences of n b/w slow and fast
        while( count <= n && fast != null){
            fast = fast.next;
            count++;
        }

        // slow will be at n-1th element 
        while(fast != null){
            slow = slow.next;
            fast = fast.next;
        }

        ListNode temp = slow.next;
        slow.next = temp.next;
        temp.next = null;

        return dummy.next;
    }
}